python-checks/src/main/java/org/sonar/python/checks/CsrfDisabledCheck.java

```diff
@@ -0,0 +1,301 @@
+    CallExpression callExpr = (CallExpression) subscriptionContext.syntaxNode();
+    boolean isFlaskFormInstantiation = Optional.ofNullable(callExpr.calleeSymbol())
+      .flatMap(tryCast(ClassSymbol.class))
+      .filter(c -> c.canBeOrExtend("flask_wtf.FlaskForm"))
+      .isPresent();
+    if (!isFlaskFormInstantiation) {
+      return;
+    }
+
+    callExpr.arguments().forEach(arg -> {
+      if (arg instanceof RegularArgument) {
+        RegularArgument regArg = (RegularArgument) arg;
+        searchForProblemsInFormInitializationArguments(regArg)
+          .ifPresent(badExpr -> subscriptionContext.addIssue(badExpr, DISABLING_CSRF_MESSAGE));
+      }
+    });
+  }
+
+  /**
+   * Attempts to find dangerous settings in a regular argument used in Flask form initialization.
+   */
+  private static Optional<Expression> searchForProblemsInFormInitializationArguments(RegularArgument regArg) {
+    String name = Optional.ofNullable(regArg.keywordArgument()).map(Name::name).orElse(null);
+    if ("csrf_enabled".equals(name) && Expressions.isFalsy(regArg.expression())) {
+      return Optional.of(regArg.expression());
+    } else if ("meta".equals(name)) {
+      return tryCast(DictionaryLiteral.class, regArg.expression())
+        .flatMap(CsrfDisabledCheck::searchForBadCsrfSettingInDictionary);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  /** Looks for <code>'csrf': False</code> and similar settings in a dictionary. */
+  private static Optional<Expression> searchForBadCsrfSettingInDictionary(DictionaryLiteral dict) {
+    return dict.elements().stream()
+      .filter(KeyValuePair.class::isInstance)
+      .map(KeyValuePair.class::cast)
+      .filter(kvp -> tryCast(StringLiteral.class, kvp.key())
+        .filter(strLit -> "csrf".equals(strLit.trimmedQuotesValue()))
+        .isPresent())
+      .findFirst()
+      .filter(kvp -> Expressions.isFalsy(kvp.value()))
+      .map(KeyValuePair::value);
+  }
+
+  private static void improperlyConfiguredFlaskApp(SubscriptionContext subscriptionContext) {
+    AssignmentStatement asgn = (AssignmentStatement) subscriptionContext.syntaxNode();
+    if (isFlaskAppInstantiation(asgn.assignedValue())) {
+      boolean isCsrfEnabledInThisFile = asgn.lhsExpressions().stream()
+        .flatMap(exprList -> exprList.expressions().stream())
+        .findFirst()
+        .flatMap(tryCast(Name.class))
+        .flatMap(app -> Optional.of(app)
+          .map(Name::symbol)
+          .map(Symbol::usages)
+          .flatMap(usages -> usages.stream().filter(CsrfDisabledCheck::isWithinCsrfEnablingStatement).findFirst()))
+        .isPresent();
+      if (!isCsrfEnabledInThisFile) {
+        subscriptionContext.addIssue(asgn.assignedValue(), CSRFPROTECT_MISSING_MESSAGE);
+      }
+    }
+  }
+
+  /** Checks that an expression is some kind of <code>Flask(...)</code> constructor invocation. */
+  private static boolean isFlaskAppInstantiation(Expression expr) {
```

Comment:
Admittedly, "tryCast" is a rather bad name: it might invoke the impression that it attempts something like "testing by casting", which sounds as if one would attempt to cast something, and then catch an exception. That's of course not the case: the actual implementation uses `instanceof` and the `cast`, it's not trying to forcibly cast anything.